presto-spark/src/main/java/com/facebook/presto/spark/planner/SparkRddPlanner.java

```diff
@@ -0,0 +1,254 @@
+        private final PrestoSparkTaskCompilerFactory compilerFactory;
+        private final int hashPartitionCount;
+        private final CollectionAccumulator<byte[]> taskStatsCollector;
+        private final TableWriteInfo tableWriteInfo;
+
+        private RddFactory(
+                Session session,
+                JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec,
+                JavaSparkContext sparkContext,
+                PrestoSparkTaskCompilerFactory taskCompilerFactory,
+                int hashPartitionCount,
+                CollectionAccumulator<byte[]> taskStatsCollector,
+                TableWriteInfo tableWriteInfo)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.sparkTaskRequestJsonCodec = requireNonNull(sparkTaskRequestJsonCodec, "sparkTaskRequestJsonCodec is null");
+            this.sparkContext = requireNonNull(sparkContext, "sparkContext is null");
+            this.compilerFactory = requireNonNull(taskCompilerFactory, "taskCompilerFactory is null");
+            this.hashPartitionCount = hashPartitionCount;
+            this.taskStatsCollector = requireNonNull(taskStatsCollector, "taskStatsCollector is null");
+            this.tableWriteInfo = requireNonNull(tableWriteInfo, "tableWriteInfo is null");
+        }
+
+        public JavaPairRDD<Integer, byte[]> createRdd(SubPlanWithTaskSources subPlan)
+        {
+            PlanFragment fragment;
+            if (subPlan.getFragment().getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+                fragment = subPlan.getFragment().withBucketToPartition(Optional.of(IntStream.range(0, hashPartitionCount).toArray()));
+            }
+            else {
+                fragment = subPlan.getFragment();
+            }
+
+            checkArgument(!fragment.getStageExecutionDescriptor().isStageGroupedExecution(), "unexpected grouped execution fragment: %s", fragment.getId());
+
+            // scans
+            List<PlanNodeId> tableScans = fragment.getTableScanSchedulingOrder();
+
+            // source stages
+            List<RemoteSourceNode> remoteSources = fragment.getRemoteSourceNodes();
+            checkArgument(tableScans.isEmpty() || remoteSources.isEmpty(), "stage has both, remote sources and table scans");
+
+            if (!tableScans.isEmpty()) {
+                checkArgument(fragment.getPartitioning().equals(SOURCE_DISTRIBUTION), "unexpected table scan partitioning: %s", fragment.getPartitioning());
+
+                List<SparkTaskDescriptor> taskRequests = subPlan.getTaskSources().stream()
+                        .flatMap(taskSource -> taskSource.getSplits().stream()
+                                .map(split -> createTaskDescriptor(
+                                        fragment,
+                                        ImmutableList.of(
+                                                new TaskSource(
```

Comment:
This will create one task request per each split, which I will rewrite. 

With N spark tasks , we should just have N `SparkTaskDescriptor`. This is because `ParallelCollectionRDD` always slice the provided collection evenly: https://github.com/apache/spark/blob/7bff2db9ed803e05a43c2d875c1dea819d81248a/core/src/main/scala/org/apache/spark/rdd/ParallelCollectionRDD.scala#L110-L117